xmtp_mls/src/identity/xmtp_id/identity.rs

```diff
@@ -0,0 +1,240 @@
+use ed25519_dalek::SigningKey;
+use openmls::{
+    credentials::{errors::BasicCredentialError, BasicCredential},
+    prelude::Credential as OpenMlsCredential,
+};
+use openmls_basic_credential::SignatureKeyPair;
+use prost::Message;
+use sha2::{Digest, Sha512};
+use thiserror::Error;
+use xmtp_id::{
+    associations::{
+        builder::{SignatureRequest, SignatureRequestBuilder, SignatureRequestError},
+        InstallationKeySignature, LegacyDelegatedSignature, MemberIdentifier,
+        RecoverableEcdsaSignature,
+    },
+    constants::INSTALLATION_KEY_SIGNATURE_CONTEXT,
+    InboxId,
+};
+use xmtp_proto::{
+    api_client::{XmtpIdentityClient, XmtpMlsClient},
+    xmtp::{
+        identity::MlsCredential,
+        message_contents::{signed_private_key, SignedPrivateKey as LegacySignedPrivateKeyProto},
+    },
+};
+use xmtp_v2::k256_helper;
+
+use crate::{
+    api::{ApiClientWrapper, GetIdentityUpdatesV2Filter, WrappedApiError},
+    configuration::CIPHERSUITE,
+};
+
+#[derive(Debug, Error)]
+pub enum IdentityError {
+    #[error(transparent)]
+    CredentialSerialization(#[from] prost::EncodeError),
+    #[error(transparent)]
+    ApiError(#[from] WrappedApiError),
+    #[error(transparent)]
+    SignatureRequestBuilder(#[from] SignatureRequestError),
+    #[error(transparent)]
+    BasicCredential(#[from] BasicCredentialError),
+    #[error("Legacy key re-use")]
+    LegacyKeyReuse,
+    #[error("Malformed legacy key: {0}")]
+    MalformedLegacyKey(String),
+    #[error("Legacy signature: {0}")]
+    LegacySignature(String),
+}
+
+pub struct Identity {
+    pub(crate) inbox_id: InboxId,
+    pub(crate) installation_keys: SignatureKeyPair,
+    pub(crate) credential: OpenMlsCredential,
+    pub(crate) signature_request: Option<SignatureRequest>,
+}
+
+impl Identity {
+    fn is_ready(&self) -> bool {
+        self.signature_request.is_none()
+    }
+
+    pub(crate) async fn create_to_be_signed<ApiClient: XmtpMlsClient + XmtpIdentityClient>(
+        inbox_id: InboxId,
+        nonce: u64,
+        account_address: String,
+        api_client: ApiClientWrapper<ApiClient>,
+    ) -> Result<Self, IdentityError> {
+        let signature_keys = SignatureKeyPair::new(CIPHERSUITE.signature_algorithm())?;
+        let installation_public_key = signature_keys.public();
+        let member_identifier: MemberIdentifier = account_address.into();
+
+        let builder = SignatureRequestBuilder::new(inbox_id);
+
+        if !Self::is_existing_inbox(inbox_id, api_client).await? {
```

Comment:
Maybe we want to do this check as part of the IdentityStrategy, and just pass in the result. I guess it'll depend on whether we need this information in the strategy's business logic too. We'd only want to do this expensive call once.

Reply:
removed `is_existing_inbox` altogether and refactored the `create_to_be_signed`, `create_from_legacy`. 
Will create a quick PR aggregate both functions so that we have LegacyKey -> Address fallback flow, as you suggested.